File: Binary Search/33. Search in Rotated Sorted Array/Solution.java
class Solution {
    public int search(int[] nums, int target) {
        // N : len(nums)
        // Time Complexity: O(N)
        // Space Complexity: O(1)
        // int n = nums.length;
        // for (int i = 0; i < n; i++) {
        //     if (nums[i] == target) return i;
        // }
        // return -1;


        // N : len(nums)
        // Time Complexity: O(logN)
        // Space Complexity: O(1)
        int n = nums.length;
        int l = 0;
        int r = n - 1;
        while (l <= r) {
            int mid = (r + l) / 2;
            if (nums[mid] == target) return mid;
            if (nums[l] <= nums[mid]) {
                if (target >= nums[l] && target <= nums[mid]) r = mid - 1;
                else l = mid + 1;
            } else {
                if (target >= nums[mid] && target <= nums[r]) l = mid + 1;
                else r = mid - 1;
            }
        }
        return -1;
    }
}

File: Binary Search/33. Search in Rotated Sorted Array/Solution.py
from typing import List

class Solution:
    def search(self, nums: List[int], target: int) -> int:
        # N : len(nums)
        # Time Complexity: O(N)
        # Space Complexity: O(1)
        # n = len(nums)
        # for i in range(n):
        #     if nums[i] == target:
        #         return i
        # return -1
        

        # N : len(nums)
        # Time Complexity: O(logN)
        # Space Complexity: O(1)
        n = len(nums)
        l, r = 0, n - 1
        while l <= r:
            mid = (r + l) // 2
            if target == nums[mid]:
                    return mid
            if nums[l] <= nums[mid]:
                if nums[l] <= target <= nums[mid]:
                    r = mid - 1
                else:
                    l = mid + 1
            elif nums[r] >= nums[mid]:
                if nums[mid] <= target <= nums[r]:
                    l = mid + 1
                else:
                    r = mid - 1
        return -1